CUDADataFormats/TrackingRecHit/interface/TrackingRecHit2DHeterogeneous.h

```diff
@@ -95,39 +101,100 @@ class TrackingRecHit2DHeterogeneous {
 
   uint32_t const* m_hitsModuleStart;  // needed for legacy, this is on GPU!
 
-  uint32_t m_nMaxModules;
   // needed as kernel params...
   PhiBinner* m_phiBinner;
-  PhiBinner::index_type* m_phiBinnerStorage;
+  typename PhiBinner::index_type* m_phiBinnerStorage;
   uint32_t* m_hitsLayerStart;
   int16_t* m_iphi;
 };
 
-using TrackingRecHit2DGPU = TrackingRecHit2DHeterogeneous<cms::cudacompat::GPUTraits>;
-using TrackingRecHit2DCPU = TrackingRecHit2DHeterogeneous<cms::cudacompat::CPUTraits>;
-using TrackingRecHit2DHost = TrackingRecHit2DHeterogeneous<cms::cudacompat::HostTraits>;
+// TrackingRecHit2DGPU/CPU/Host derived classes workaround to have partial specialization.
+// This is neeeded since one could not partially specialize a single method without specializing the whole class.
+// So simply having an alias such as
+//
+// template <typename TrackerTraits>
+// using TrackingRecHit2DGPUT = TrackingRecHit2DHeterogeneousT<cms::cudacompat::GPUTraits, TrackerTraits>;
+//
+// wouldn't work (giving some "invalid use of incomplete type" error). The alternative would be to have the whole class specialized. But this would mean we need
+// to rewrite every method (specialzed don't see primary methods). This is overcome with inheritance.
+// Thus with this workaround, i.e. using inheritance+specialization, we keep under control code duplication.
+// Another solution would be to have an extra accessor struct in the class to partially specialize only that
+// (as described here https://stackoverflow.com/questions/165101/invalid-use-of-incomplete-type-error-with-partial-template-specialization)
+// but it seems to me more messy than this.
+// The same reasoning applies to CAHitNtupletGeneratorKernels.
+
+template <typename Traits, typename TrackerTraits>
+class TrackingRecHit2DGPUBaseT : public TrackingRecHit2DHeterogeneousT<cms::cudacompat::GPUTraits, TrackerTraits> {};
+
+template <typename Traits, typename TrackerTraits>
+class TrackingRecHit2DCPUBaseT : public TrackingRecHit2DHeterogeneousT<cms::cudacompat::CPUTraits, TrackerTraits> {};
+
+template <typename Traits, typename TrackerTraits>
+class TrackingRecHit2DHostBaseT : public TrackingRecHit2DHeterogeneousT<cms::cudacompat::HostTraits, TrackerTraits> {};
```

Comment:
This looks overly complicated: you don't need to inherit keeping both template parameters and then always specialise the first one -- you can just keep only the second one in the inherited class:
```c++
// Inherit and overload only what we need to overload
// GPU
template <typename TrackerTraits>
class TrackingRecHit2DGPUT
    : public TrackingRecHit2DHeterogeneousT<cms::cudacompat::GPUTraits, TrackerTraits> {
public:
  using TrackingRecHit2DHeterogeneousT<cms::cudacompat::GPUTraits, TrackerTraits>::TrackingRecHit2DHeterogeneousT;
  cms::cuda::host::unique_ptr<float[]> localCoordToHostAsync(cudaStream_t stream) const;
  cms::cuda::host::unique_ptr<uint32_t[]> hitsModuleStartToHostAsync(cudaStream_t stream) const;
  cms::cuda::host::unique_ptr<uint16_t[]> store16ToHostAsync(cudaStream_t stream) const;
  cms::cuda::host::unique_ptr<float[]> store32ToHostAsync(cudaStream_t stream) const;
};
```

and remove the aliases down in the file.

Reply:
Indeed you are right, I've followed your suggestion and everything is smooth.